### PR TITLE
[WIP] Knockback

### PR DESCRIPTION
**Pull request by Dumbeldor**
Linked with #5679 

Video :

https://www.youtube.com/watch?v=8t-hVu91IXI

### CODE REVIEW

**Comment by nerzhul** [comment]
dupe

**Comment by nerzhul** [comment]
Linked with #5679

**Comment by paramat** [comment]
I'm wondering why you opened a new PR, and why it already has conflicts?

**Comment by Fixer-007** [comment]
Please add coherent description of this PR.

**Comment by Dumbeldor** [comment]
Yesterday i fail my rebase... I dont know why but my first PR is closed after my push :D

**Comment by paramat** [comment]
Ok no problem.
See https://github.com/minetest/minetest/pull/3208#issuecomment-334113859 can this be a more general 'throw object' feature so that it can be used for more than just knockback?

**Comment by Dumbeldor** [comment]
Yes @paramat it is a good idea !

File: src/content_cao.cpp [hunk 1]
```diff
@@ -830,7 +830,8 @@ void GenericCAO::step(float dtime, ClientEnvironment *env)
 	} else {
 		v3f lastpos = pos_translator.vect_show;
 
-		if(m_prop.physical) {
+		if (m_prop.physical)
+		{
```
Comment (nerzhul) [inline src/content_cao.cpp]: code style

File: src/content_sao.cpp [hunk 2]
```diff
@@ -1245,7 +1246,7 @@ int PlayerSAO::punch(v3f dir,
 					m_knockback_timer = time_reset_knockback;
 					m_knockback = true;
 					m_env->getGameDef()->SendKnockBack(m_peer_id, direction, m_knockback_timer, getId());
-					m_knockback_timer *= 5;
+					m_knockback_timer;
```
Comment (nerzhul) [inline src/content_sao.cpp]: dead code ?

File: src/content_sao.cpp [hunk 3]
```diff
@@ -1024,7 +1024,8 @@ void PlayerSAO::step(float dtime, bool send_recommended)
 	}
 
 	// If the object is attached client-side, don't waste bandwidth sending its position to clients
-	if (m_position_not_sent && !isAttached() && !m_knockback) {
+	if (m_position_not_sent && !isAttached() && !m_knockback)
```
Comment (nerzhul) [inline src/content_sao.cpp]: code style ?

File: src/localplayer.cpp [hunk 4]
```diff
@@ -684,9 +688,10 @@ void LocalPlayer::applyControl(float dtime, Environment *env)
 	float slip_factor = getSlipFactor(env, speedH);
 	// Accelerate to target speed with maximum increment
 	accelerateHorizontal(speedH * physics_override_speed,
-			incH * physics_override_speed * slip_factor);
+						 incH * physics_override_speed * slip_factor);
 	accelerateVertical(speedV * physics_override_speed,
-			incV * physics_override_speed);
+					   incV * physics_override_speed);
```
Comment (nerzhul) [inline src/localplayer.cpp]: indent problem

File: src/content_cao.cpp [hunk 5]
```diff
@@ -830,8 +830,20 @@ void GenericCAO::step(float dtime, ClientEnvironment *env)
 	} else {
 		v3f lastpos = pos_translator.vect_show;
 
-		if(m_prop.physical)
+		if (m_prop.physical)
 		{
```
Comment (nerzhul) [inline src/content_cao.cpp]: brace

File: src/server.cpp [hunk 6]
```diff
@@ -1849,6 +1867,28 @@ void Server::SendMovePlayer(session_t peer_id)
 	Send(&pkt);
 }
 
+void Server::SendKnockBack(const u16 peer_id, const v3f &direction, const f32 time_knockback,
```
Comment (nerzhul) [inline src/server.cpp]: const session_t peer_id

File: src/server.cpp [hunk 7]
```diff
@@ -1279,6 +1279,24 @@ void Server::SetBlocksNotSent(std::map<v3s16, MapBlock *>& block)
 	m_clients.unlock();
 }
 
+void Server::SendPacketInRange(const v3f &pos, u32 range, NetworkPacket *pkt)
+{
+	std::vector<u16> clients = m_clients.getClientIDs();
+
+	for (std::vector<u16>::iterator i = clients.begin(); i != clients.end(); ++i) {
```
Comment (nerzhul) [inline src/server.cpp]: use a c++11 range based for loop

File: src/content_sao.cpp (hunk 2 above)
Comment (Dumbeldor) [reply src/content_sao.cpp]: Oh yes i forget :D

**Comment by paramat** [comment]
Does this now have a ballistic motion with initial velocity in the direction of the punch, as requested?

**Comment by NewbProgrammer101** [comment]
The knockback is pushing the velocity of the player model too far. Decrease velocity please. I can't bear to look at how far the player's model can be pushed from one punch.

**Comment by lhofhansl** [comment]
Perhaps make the strength of the effect configurable, with a lower default.

**Comment by paramat** [comment]
If this still has the same knockback motion as the original PR it is still unacceptable. it needs to be a ballistic arc (physically correct motion with acceleration = gravity), with initial velocity equal to punch direction (or fallback to a reasonable angle).

**Comment by Fixer-007** [comment]
>If this still has the same knockback motion https://www.youtube.com/watch?v=8t-hVu91IXI as the original PR it is still unacceptable.
knockback looks extremely powerful, it needs to be configurable.

**Comment by Dumbeldor** [comment]
You can configure the power of the knockback in minetest.conf like this : 
`knockback_horizontal=200
knockback_vertical=100
knockback_default_duration = 0.4
`
@paramat i change, i applique a ballistic motion.
What do you think @nerzhul ?

**Comment by nerzhul** [comment]
Please make it configurable via lua api

**Comment by paramat** [comment]
Strength of knockback should be defined by damage inflicted, it doesn't need a setting.

**Comment by DS-Minetest** [comment]
>Strength of knockback (initial velocity) should be defined by damage inflicted

Nah, stabbing with a dagger can make much damage but shouldn't give much knockback while a very heavy weapon should make more knockback. I suggest instead a new `knockback` `tool_capability` or `damage_group`.

**Comment by paramat** [comment]
@Dumbeldor PR is neglected and the implementation is still bad, does it need adoption?

**Comment by paramat** [comment]
Neglected, closing for now.

**Pull request closed without merging**